src/Tools/Source/RunTests/Program.cs

```diff
@@ -41,7 +42,34 @@ internal static int Main(string[] args)
                 cts.Cancel();
             };
 
-            return RunCore(options, cts.Token).GetAwaiter().GetResult();
+            return Run(options, cts.Token).GetAwaiter().GetResult();
+        }
+
+        private static async Task<int> Run(Options options, CancellationToken cancellationToken)
+        {
+            if (options.Timeout == null)
+            {
+                return await RunCore(options, cancellationToken);
+            }
+
+            var timeoutTask = Task.Delay(options.Timeout.Value);
+            var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            var runTask = RunCore(options, cts.Token);
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return ExitFailure;
+            }
+
+            var finishedTask = await Task.WhenAny(timeoutTask, runTask);
+            if (finishedTask == timeoutTask)
+            {
+                HandleTimeout(options);
+                cts.Cancel();
```

Comment:
So this will pass the cancellation, but do we need to wiat for runTask? Otherwise we might bail and our process go away prior to actually killing child tasks.